mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java

```diff
@@ -125,13 +125,14 @@ public FieldDeserializer primitive(PrimitiveType type, ObjectInspectorPair pair)
 
     @Override
     public FieldDeserializer struct(StructType type, ObjectInspectorPair pair, List<FieldDeserializer> deserializers) {
+      GenericRecord template = type != null ? GenericRecord.create(type) : null;
       return o -> {
         if (o == null) {
           return null;
         }
 
         List<Object> data = ((StructObjectInspector) pair.sourceInspector()).getStructFieldsDataAsList(o);
-        Record result = GenericRecord.create(type);
+        Record result = template.copy();
```

Comment:
Can this result in NPE?

Reply:
Then there were an NPE with the original code as well. The template is only null, when the type is null.
```
  private GenericRecord(StructType struct) {
    this.struct = struct;
    this.size = struct.fields().size();         <-- NPE here when the struct is null
    this.values = new Object[size];
    this.nameToPos = NAME_MAP_CACHE.get(struct);
  }
```

Reply:
Indeed, the `type` should never be null here because the `SchemaWithPartnerVistor` guarantee it. See: https://github.com/apache/iceberg/blob/90225d6c9413016d611e2ce5eff37db1bc1b4fc5/core/src/main/java/org/apache/iceberg/schema/SchemaWithPartnerVisitor.java#L52-L61

I think we can just add a `Precondition.checkNotNull` in the line128 instead of checking the null case in a `if-else` block.

Reply:
So here we use `GenericRecord(GenericRecord toCopy)` constructor rather than `GenericRecord(StructType struct)` to eliminate the access to `NAME_MAP_CACHE` ?  I'm afraid the further PR will easily fallback to use the other because this tiny different is hard for the next developer and reviewer to notice.  Is possible to add few doc ?

Reply:
Done

Reply:
There were an issue with positional deletes where the row was not present in the file. The code below still created reader for the empty struct.
```
      Types.StructType struct = iType != null ? iType.asStructType() : null;
      return visitor.struct(struct, group, visitFields(struct, group, visitor));
```

Changed it to this, let's see if it breaks something:
```
      return iType != null ?
              visitor.struct(iType.asStructType(), group, visitFields(iType.asStructType(), group, visitor)) : null;
```

Reply:
Had to revert the changes.
In the Deserializer we do not have null struct so I added the `Preconditions.checkNotNull()` check, but I had to revert to the null checks because the Spark reader expects the visitors/readers to handle null structs when the readers are created.

I had failures with `TestSparkReadProjection.testListOfStructsProjection()` and `TestGenericAppenderFactory.testPosDeleteWriterWithRowSchema()`